homeassistant/components/sensor/citybikes.py

```diff
@@ -250,13 +251,19 @@ def state(self):
         """Return the state of the sensor."""
         return self._station_data.get(ATTR_FREE_BIKES, STATE_UNKNOWN)
 
+    @property
+    def entity_id(self):
+        """Return the entity ID."""
+        return "{}.{}_{}".format(PARENT_DOMAIN,
```

Comment:
Remove the property and define an instance attribute `self.entity_id` in init instead. This isn't doing any I/O, correct?

Reply:
And it is recommended to use `generate_entity_id` as @OttoWinter says.

Only entities that defines a unique ID get registered in the entity registry. So id conflict can still occur for all other entities.

Reply:
For some reason, calling `generate_entity_id` gets stuck. I think it has something to do with the fact that the platform setup is async, but the call to `generate_entity_id` is performed in a constructor, which is not async. From looking at the code of other platforms and components, it seems that only platforms that aren't async use `generate_entity_id`, while others that need the functionality, simply create the ID on their own - which is what I've done here (after all, the logic for creating the ID is pretty straightforward, and shouldn't result in any duplicates).

Reply:
Then you can just call `async_generate_entity_id` ;)

Reply:
I tried that, that doesn't work either... :-/...
I'm supposed to `yield from` it, am I not? This is what I tried in the constructor:
`self.entity_id = yield from async_generate_entity_id(ENTITY_ID_FORMAT, uid, hass=hass)`

It results in an exception:
```
Traceback (most recent call last):
  File "/srv/homeassistant/lib/python3.6/site-packages/homeassistant/helpers/entity_platform.py", line 84, in async_setup
    SLOW_SETUP_MAX_WAIT, loop=hass.loop)
  File "/usr/local/lib/python3.6/asyncio/tasks.py", line 358, in wait_for
    return fut.result()
  File "/usr/local/lib/python3.6/asyncio/futures.py", line 245, in result
    raise self._exception
  File "/usr/local/lib/python3.6/asyncio/tasks.py", line 180, in _step
    result = coro.send(None)
  File "/srv/homeassistant/lib/python3.6/site-packages/homeassistant/components/sensor/citybikes.py", line 172, in async_setup_platform
    devices.append(CityBikesStation(hass, network, station_id, name))
TypeError: __init__() should return None, not 'generator'
```

And when using the normal version, as I mentioned before, it just hangs inside the execution of `generate_entity_id`...

Reply:
Yeah, sorry for unclear comments. You can't `yield from` in the constructor. So you have to do generate the entity_id in `async_setup_platform` and pass the id as an argument to the constructor.

Reply:
OK, after more experimentations, tried calling `async_generate_entity_id` directly (without `yield from`), and it worked. I'm confused - I thought I'm supposed to `await` or to `yield from` any coroutines...

Reply:
It's not a coroutine, it's a callback.

Reply:
It seems that the naming is a bit confusing, in that case... Oh well.

Reply:
A function that is named `async_*` in home assistant means that the function needs to run in the event loop. Coroutines are marked with the `@asyncio.coroutine` annotation or defined with the python 3.5 keyword `async def`.

Reply:
I see. Thanks!